From 6f48c379d65e7d32b111327c05587811db4d5323 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Wed, 29 Oct 2014 15:54:56 -0400
Subject: [ethernet] mlx4: Add ethernet backplane autoneg device capability

Message-id: <f87c130df91b0ee3f4d47355ad45dd441b64e0eb.1414597479.git.avadai@redhat.com>
Patchwork-id: 98511
O-Subject: [Patch RHEL7.1 5/9] net/mlx4_core: Add ethernet backplane autoneg device capability
Bugzilla: 1060221
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Saeed Mahameed <saeedm@mellanox.com>

Signed-off-by: Saeed Mahameed <saeedm@mellanox.com>
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit a53e3e8c1db547981e13d1ebf24a659bd4e87710)
Signed-off-by: Amir Vadai <avadai@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/fw.c b/drivers/net/ethernet/mellanox/mlx4/fw.c
index 5b919a0..ddcd47c 100644
--- a/drivers/net/ethernet/mellanox/mlx4/fw.c
+++ b/drivers/net/ethernet/mellanox/mlx4/fw.c
@@ -140,7 +140,8 @@ static void dump_dev_cap_flags2(struct mlx4_dev *dev, u64 flags)
   [11] = "MAD DEMUX (Secure-Host) support",
   [12] = "Large cache line (>64B) CQE stride support",
   [13] = "Large cache line (>64B) EQE stride support",
-  [14] = "Ethernet protocol control support"
+  [14] = "Ethernet protocol control support",
+  [15] = "Ethernet Backplane autoneg support"
  };
  int i;
 
@@ -575,6 +576,7 @@ int mlx4_QUERY_DEV_CAP(struct mlx4_dev *dev, struct mlx4_dev_cap *dev_cap)
 #define QUERY_DEV_CAP_BMME_FLAGS_OFFSET  0x94
 #define QUERY_DEV_CAP_RSVD_LKEY_OFFSET  0x98
 #define QUERY_DEV_CAP_MAX_ICM_SZ_OFFSET  0xa0
+#define QUERY_DEV_CAP_ETH_BACKPL_OFFSET  0x9c
 #define QUERY_DEV_CAP_FW_REASSIGN_MAC  0x9d
 #define QUERY_DEV_CAP_VXLAN   0x9e
 #define QUERY_DEV_CAP_MAD_DEMUX_OFFSET  0xb0
@@ -749,6 +751,9 @@ int mlx4_QUERY_DEV_CAP(struct mlx4_dev *dev, struct mlx4_dev_cap *dev_cap)
    QUERY_DEV_CAP_BMME_FLAGS_OFFSET);
  MLX4_GET(dev_cap->reserved_lkey, outbox,
    QUERY_DEV_CAP_RSVD_LKEY_OFFSET);
+ MLX4_GET(field32, outbox, QUERY_DEV_CAP_ETH_BACKPL_OFFSET);
+ if (field32 & (1 << 0))
+  dev_cap->flags2 |= MLX4_DEV_CAP_FLAG2_ETH_BACKPL_AN_REP;
  MLX4_GET(field, outbox, QUERY_DEV_CAP_FW_REASSIGN_MAC);
  if (field & 1<<6)
   dev_cap->flags2 |= MLX4_DEV_CAP_FLAG2_REASSIGN_MAC_EN;
diff --git a/include/linux/mlx4/device.h b/include/linux/mlx4/device.h
index bfcd320..1bf8506 100644
--- a/include/linux/mlx4/device.h
+++ b/include/linux/mlx4/device.h
@@ -187,7 +187,8 @@ enum {
  MLX4_DEV_CAP_FLAG2_MAD_DEMUX  = 1LL <<  11,
  MLX4_DEV_CAP_FLAG2_CQE_STRIDE  = 1LL <<  12,
  MLX4_DEV_CAP_FLAG2_EQE_STRIDE  = 1LL <<  13,
- MLX4_DEV_CAP_FLAG2_ETH_PROT_CTRL        = 1LL <<  14
+ MLX4_DEV_CAP_FLAG2_ETH_PROT_CTRL        = 1LL <<  14,
+ MLX4_DEV_CAP_FLAG2_ETH_BACKPL_AN_REP = 1LL <<  15
 };
 
 enum {
-- 
1.7.1